gemini/reasoning-engine/tutorial_langgraph.ipynb

```diff
@@ -0,0 +1,665 @@
+    "\n",
+    "This notebook demonstrates how to build, deploy, and test a simple LangGraph application using [Reasoning Engine](https://cloud.google.com/vertex-ai/generative-ai/docs/reasoning-engine/overview) in Vertex AI. You'll learn how to combine LangGraph's workflow orchestration with the scalability of Vertex AI, which enables you to build custom generative AI applications.\n",
+    "\n",
+    "Note that the approach used in this notebook defines a [custom application template in Reasoning Engine](https://cloud.google.com/vertex-ai/generative-ai/docs/reasoning-engine/customize), which can be extended to LangChain or other orchestration frameworks. If just want to use LangChain on Vertex AI to build agentic generative AI applications, refer to the documentation for [developing with the LangChain template in Reasoning Engine](https://cloud.google.com/vertex-ai/generative-ai/docs/reasoning-engine/develop).\n",
+    "\n",
+    "This notebook covers the following steps:\n",
+    "\n",
+    "- **Define Tools**: Create custom Python functions to act as tools your AI application can use.\n",
+    "- **Define Router**: Create custom Python functions to act as tools your AI application can use.\n",
+    "- **Build a LangGraph Application**: Structure your application using LangGraph, including the Gemini model and custom tools that you define.\n",
+    "- **Local Testing**: Test your LangGraph application locally to ensure functionality.\n",
+    "- **Deploying to Vertex AI**: Seamlessly deploy your LangGraph application to Reasoning Engine for scalable execution.\n",
+    "- **Remote Testing**: Interact with your deployed application through Vertex AI, testing its functionality in a production-like environment.\n",
+    "- **Cleaning Up Resources**: Delete your deployed application on Vertex AI to avoid incurring unnecessary charges.\n",
+    "\n",
+    "By the end of this notebook, you'll have the skills and knowledge to build and deploy your own custom generative AI applications using LangGraph, Reasoning Engine, and Vertex AI."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "id": "61RBz8LLbxCR"
+   },
+   "source": [
+    "## Get started"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "id": "No17Cw5hgx12"
+   },
+   "source": [
+    "### Install Vertex AI SDK and other required packages\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 1,
+   "metadata": {
+    "id": "tFy3H3aPgx12"
+   },
+   "outputs": [],
+   "source": [
+    "# %pip install --upgrade --user --quiet \\\n",
+    "#     \"google-cloud-aiplatform[langchain,reasoningengine]\" \\\n",
+    "#     cloudpickle==3.0.0 \\\n",
+    "#     pydantic==2.7.4 \\\n",
+    "#     langgraph"
```

Comment:
```suggestion
    "%pip install --upgrade --user --quiet \\\n",
    "    \"google-cloud-aiplatform[langchain,reasoningengine]\" \\\n",
    "    cloudpickle==3.0.0 \\\n",
    "    pydantic==2.7.4 \\\n",
    "    langgraph"
```

Is there a reason these were commented out?

Reply:
Also, later when deploying on reasoning engine, you add `httpx` should that be here as well?

Reply:
Accidentally left these commented when testing 😊 . Fixed!

Reply:
`httpx` is only really needed in the remote serving environment, but I'll add it to the top of the notebook for consistency.